File: src/transformer/rope.rs
use std::collections::HashMap;

use crate::common::num_traits::FromNumber;
use serde_json::Value;

#[inline]
fn inv_freqs(dim: usize, theta: f32) -> Vec<f32> {
    debug_assert!(dim % 2 == 0, "RoPE head_dim must be even");
    (0..dim)
        .step_by(2)
        .map(|i| {
            let p = i as f32 / dim as f32;
            theta.powf(p).recip()
        })
        .collect()
}

#[derive(Debug, Clone, Copy, PartialEq)]
struct YarnScaling {
    factor: f32,
    original_max_position_embeddings: usize,
    attention_factor: f32,
    beta_fast: f32,
    beta_slow: f32,
}

fn parse_rope_scaling(
    rope_scaling: Option<&HashMap<String, Value>>,
    default_original_max_position_embeddings: usize,
) -> Option<YarnScaling> {
    let rope_scaling = rope_scaling?;

    let rope_type = rope_scaling
        .get("rope_type")
        .or_else(|| rope_scaling.get("type"))?
        .as_str()?
        .to_ascii_lowercase();

    if rope_type != "yarn" {
        return None;
    }

    let factor = rope_scaling
        .get("factor")
        .and_then(value_to_f32)
        .filter(|factor| *factor > 0.0)
        .unwrap_or(1.0);

    let original_max_position_embeddings = rope_scaling
        .get("original_max_position_embeddings")
        .and_then(value_to_usize)
        .unwrap_or(default_original_max_position_embeddings);

    let beta_fast = rope_scaling
        .get("beta_fast")
        .and_then(value_to_f32)
        .unwrap_or(32.0);

    let beta_slow = rope_scaling
        .get("beta_slow")
        .and_then(value_to_f32)
        .unwrap_or(1.0);

    let attention_factor = rope_scaling
        .get("attention_factor")
        .or_else(|| rope_scaling.get("attn_factor"))
        .and_then(value_to_f32)
        .unwrap_or_else(|| 0.1 * factor.ln() + 1.0);

    Some(YarnScaling {
        factor,
        original_max_position_embeddings,
        attention_factor,
        beta_fast,
        beta_slow,
    })
}

fn value_to_f32(value: &Value) -> Option<f32> {
    match value {
        Value::Number(number) => number.as_f64().map(|v| v as f32),
        Value::String(text) => text.parse::<f32>().ok(),
        _ => None,
    }
}

fn value_to_usize(value: &Value) -> Option<usize> {
    match value {
        Value::Number(number) => number.as_u64().map(|v| v as usize),
        Value::String(text) => text.parse::<usize>().ok(),
        _ => None,
    }
}

fn apply_yarn_scaling(
    inv_freqs: &mut [f32],
    rotary_dim: usize,
    theta: f32,
    yarn: &YarnScaling,
) {
    if yarn.factor <= 1.0 {
        return;
    }

    let rotary_pairs = rotary_dim / 2;
    if rotary_pairs == 0 {
        return;
    }

    let low_rot = yarn.beta_fast.max(1.0).round() as usize;
    let high_rot = yarn.beta_slow.max(1.0).round() as usize;
    let (low, high) = yarn_find_correction_range(
        low_rot,
        high_rot,
        rotary_dim,
        theta,
        yarn.original_max_position_embeddings,
    );
    let ramp = yarn_linear_ramp_mask(low as f32, high as f32, rotary_pairs);
    let inv_freq_extrapolation = inv_freqs.to_vec();
    let inv_freq_interpolation: Vec<f32> = inv_freq_extrapolation
        .iter()
        .map(|freq| *freq / yarn.factor)
        .collect();

    for i in 0..rotary_pairs {
        let inv_freq_mask = 1.0 - ramp[i];
        inv_freqs[i] = inv_freq_interpolation[i] * (1.0 - inv_freq_mask)
            + inv_freq_extrapolation[i] * inv_freq_mask;
    }
}

fn yarn_find_correction_dim(
    num_rotations: usize,
    dim: usize,
    base: f32,
    max_position_embeddings: usize,
) -> f32 {
    (dim as f32
        * ((max_position_embeddings as f32) / (num_rotations as f32 * 2.0 * core::f32::consts::PI))
            .ln())
        / (2.0 * base.ln())
}

fn yarn_find_correction_range(
    low_rot: usize,
    high_rot: usize,
    dim: usize,
    base: f32,
    max_position_embeddings: usize,
) -> (usize, usize) {
    let low = yarn_find_correction_dim(low_rot, dim, base, max_position_embeddings).floor();
    let high = yarn_find_correction_dim(high_rot, dim, base, max_position_embeddings).ceil();
    let low = low.max(0.0) as usize;
    let high = high.min((dim.saturating_sub(1)) as f32) as usize;
    (low, high)
}

fn yarn_linear_ramp_mask(low: f32, high: f32, dim: usize) -> Vec<f32> {
    let high = if (low - high).abs() < f32::EPSILON {
        high + 0.001
    } else {
        high
    };

    (0..dim)
        .map(|i| {
            let linear = (i as f32 - low) / (high - low);
            linear.clamp(0.0, 1.0)
        })
        .collect()
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct RotaryEmbedding {
    pub head_dim: usize,
    pub rotary_dim: usize,
    pub max_sequence_length: usize,
    pub theta: f32,
    pub attention_scaling: f32,
    yarn_scaling: Option<YarnScaling>,
}

impl RotaryEmbedding {
    pub fn new(
        head_dim: usize,
        rotary_dim: usize,
        max_sequence_length: usize,
        theta: f32,
        rope_scaling: Option<HashMap<String, Value>>,
    ) -> Self {
        let yarn = parse_rope_scaling(rope_scaling.as_ref(), max_sequence_length);
        let attention_scaling = yarn.map(|y| y.attention_factor).unwrap_or(1.0);
        Self {
            head_dim,
            rotary_dim: rotary_dim.min(head_dim),
            max_sequence_length,
            theta,
            attention_scaling,
            yarn_scaling: yarn,
        }
    }

    pub fn forward<T: FromNumber>(&self) -> Vec<T> {
        debug_assert!(self.head_dim % 2 == 0, "RoPE head_dim must be even");
        debug_assert!(
            self.rotary_dim <= self.head_dim,
            "rotary_dim must not exceed head_dim"
        );
        debug_assert!(self.rotary_dim % 2 == 0, "RoPE rotary_dim must be even");

        let rotary_pairs = self.rotary_dim / 2;
        let mut inv = inv_freqs(self.rotary_dim, self.theta);
        if let Some(yarn) = &self.yarn_scaling {
            apply_yarn_scaling(
                &mut inv,
                self.rotary_dim,
                self.theta,
                &yarn,
            );
        }
        let mut out = Vec::with_capacity(self.max_sequence_length * self.head_dim);

        for pos in 0..self.max_sequence_length {
            let t = pos as f32;

            for &inv_f in &inv {
                let angle = t * inv_f;
                out.push(T::from_f32(angle.cos() * self.attention_scaling));
                out.push(T::from_f32(angle.sin() * self.attention_scaling));
            }

            for _ in rotary_pairs..(self.head_dim / 2) {
                out.push(T::from_f32(1.0));
                out.push(T::from_f32(0.0));
            }
        }

        out
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_freqs() {
        let rope = RotaryEmbedding::new(64, 64, 512, 10000.0, None);
        let freqs = rope.forward::<f32>();
        let dim = 64;
        let max_len = 512;
        assert_eq!(freqs.len(), dim * max_len);
        // position 0 should be cos=1, sin=0 for all pairs
        for i in (0..dim).step_by(2) {
            assert!((freqs[i] - 1.0).abs() < 1e-6);
            assert!(freqs[i + 1].abs() < 1e-6);
        }
    }

    #[test]
    fn test_partial_rotary_embedding_identity_tail() {
        let rope = RotaryEmbedding::new(8, 4, 2, 10_000.0, None);
        let freqs = rope.forward::<f32>();

        assert_eq!(freqs.len(), 16);

        assert!((freqs[0] - 1.0).abs() < 1e-6);
        assert!(freqs[1].abs() < 1e-6);
        assert!((freqs[2] - 1.0).abs() < 1e-6);
        assert!(freqs[3].abs() < 1e-6);

        assert!((freqs[4] - 1.0).abs() < 1e-6);
        assert!(freqs[5].abs() < 1e-6);
        assert!((freqs[6] - 1.0).abs() < 1e-6);
        assert!(freqs[7].abs() < 1e-6);

        let pos1 = 8;
        assert!((freqs[pos1] - 1.0f32.cos()).abs() < 1e-6);
        assert!((freqs[pos1 + 1] - 1.0f32.sin()).abs() < 1e-6);
        assert!((freqs[pos1 + 2] - 0.01f32.cos()).abs() < 1e-6);
        assert!((freqs[pos1 + 3] - 0.01f32.sin()).abs() < 1e-6);
        assert!((freqs[pos1 + 4] - 1.0).abs() < 1e-6);
        assert!((freqs[pos1 + 5] - 0.0).abs() < 1e-6);
        assert!((freqs[pos1 + 6] - 1.0).abs() < 1e-6);
        assert!((freqs[pos1 + 7] - 0.0).abs() < 1e-6);
    }

    #[test]
    fn test_yarn_scaling_uses_attention_factor() {
        let mut rope_scaling = HashMap::new();
        rope_scaling.insert("rope_type".to_string(), json!("yarn"));
        rope_scaling.insert("factor".to_string(), json!(4.0));
        rope_scaling.insert(
            "original_max_position_embeddings".to_string(),
            json!(2usize),
        );
        rope_scaling.insert("attention_factor".to_string(), json!(1.25));

        let rope = RotaryEmbedding::new(8, 8, 16, 10_000.0, Some(rope_scaling));
        assert!((rope.attention_scaling - 1.25).abs() < 1e-6);

        let freqs = rope.forward::<f32>();
        let base = RotaryEmbedding::new(8, 8, 16, 10_000.0, None).forward::<f32>();

        assert_eq!(freqs.len(), base.len());
        assert!((freqs[0] - 1.25).abs() < 1e-6);
        assert!(freqs
            .iter()
            .zip(base.iter())
            .any(|(lhs, rhs)| (lhs - rhs).abs() > 1e-6));
    }
}

File: tests/reference/hf/rope.py
#!/usr/bin/env python3
"""Hugging Face style RoPE reference implementation.

This script is intentionally dependency-free so it can be used as a small
oracle for Rust alignment tests.

Input format:
{
  "head_dim": 8,
  "rotary_dim": 4,
  "max_sequence_length": 2,
  "theta": 10000.0,
  "rope_scaling": {... optional ...}
}

Output format:
{
  "head_dim": 8,
  "rotary_dim": 4,
  "max_sequence_length": 2,
  "theta": 10000.0,
  "attention_scaling": 1.0,
  "values": [...]
}
"""

from __future__ import annotations

import argparse
import json
import math
from pathlib import Path
from typing import Any


def value_to_float(value: Any) -> float | None:
    if isinstance(value, bool):
        return None
    if isinstance(value, (int, float)):
        return float(value)
    if isinstance(value, str):
        try:
            return float(value)
        except ValueError:
            return None
    return None


def value_to_int(value: Any) -> int | None:
    if isinstance(value, bool):
        return None
    if isinstance(value, int):
        return value
    if isinstance(value, float) and value.is_integer():
        return int(value)
    if isinstance(value, str):
        try:
            return int(value)
        except ValueError:
            return None
    return None


def inv_freqs(dim: int, theta: float) -> list[float]:
    if dim % 2 != 0:
        raise ValueError("RoPE head_dim must be even")
    return [theta ** (-(i / dim)) for i in range(0, dim, 2)]


def yarn_find_correction_dim(
    num_rotations: int,
    dim: int,
    base: float,
    max_position_embeddings: int,
) -> float:
    return (
        dim
        * math.log(max_position_embeddings / (num_rotations * 2.0 * math.pi))
        / (2.0 * math.log(base))
    )


def yarn_find_correction_range(
    low_rot: int,
    high_rot: int,
    dim: int,
    base: float,
    max_position_embeddings: int,
) -> tuple[int, int]:
    low = math.floor(yarn_find_correction_dim(low_rot, dim, base, max_position_embeddings))
    high = math.ceil(yarn_find_correction_dim(high_rot, dim, base, max_position_embeddings))
    low = max(0, int(low))
    high = min(int(high), max(dim - 1, 0))
    return low, high


def yarn_linear_ramp_mask(low: float, high: float, dim: int) -> list[float]:
    if abs(low - high) < 1e-12:
        high += 0.001

    out: list[float] = []
    for i in range(dim):
        linear = (i - low) / (high - low)
        out.append(min(1.0, max(0.0, linear)))
    return out


def parse_rope_scaling(
    rope_scaling: dict[str, Any] | None,
    default_original_max_position_embeddings: int,
) -> dict[str, float | int] | None:
    if not rope_scaling:
        return None

    rope_type = rope_scaling.get("rope_type", rope_scaling.get("type"))
    if not isinstance(rope_type, str) or rope_type.lower() != "yarn":
        return None

    factor = value_to_float(rope_scaling.get("factor"))
    if factor is None or factor <= 0.0:
        factor = 1.0

    original_max_position_embeddings = value_to_int(
        rope_scaling.get("original_max_position_embeddings")
    )
    if original_max_position_embeddings is None:
        original_max_position_embeddings = default_original_max_position_embeddings

    beta_fast = value_to_float(rope_scaling.get("beta_fast"))
    if beta_fast is None:
        beta_fast = 32.0

    beta_slow = value_to_float(rope_scaling.get("beta_slow"))
    if beta_slow is None:
        beta_slow = 1.0

    attention_factor = value_to_float(rope_scaling.get("attention_factor"))
    if attention_factor is None:
        attention_factor = value_to_float(rope_scaling.get("attn_factor"))
    if attention_factor is None:
        attention_factor = 0.1 * math.log(factor) + 1.0

    return {
        "factor": factor,
        "original_max_position_embeddings": original_max_position_embeddings,
        "attention_factor": attention_factor,
        "beta_fast": beta_fast,
        "beta_slow": beta_slow,
    }


def apply_yarn_scaling(
    inv_freqs_values: list[float],
    rotary_dim: int,
    theta: float,
    yarn: dict[str, float | int],
) -> None:
    factor = float(yarn["factor"])
    if factor <= 1.0:
        return

    rotary_pairs = rotary_dim // 2
    if rotary_pairs == 0:
        return

    # Rust's `round()` rounds half-away-from-zero. These values are positive,
    # so `floor(x + 0.5)` matches the same result.
    low_rot = max(1, int(math.floor(float(yarn["beta_fast"]) + 0.5)))
    high_rot = max(1, int(math.floor(float(yarn["beta_slow"]) + 0.5)))
    low, high = yarn_find_correction_range(
        low_rot,
        high_rot,
        rotary_dim,
        theta,
        int(yarn["original_max_position_embeddings"]),
    )
    ramp = yarn_linear_ramp_mask(float(low), float(high), rotary_pairs)
    inv_freq_extrapolation = list(inv_freqs_values)
    inv_freq_interpolation = [freq / factor for freq in inv_freq_extrapolation]

    for i in range(rotary_pairs):
        inv_freq_mask = 1.0 - ramp[i]
        inv_freqs_values[i] = (
            inv_freq_interpolation[i] * (1.0 - inv_freq_mask)
            + inv_freq_extrapolation[i] * inv_freq_mask
        )


def generate_rotary_embedding(case: dict[str, Any]) -> dict[str, Any]:
    head_dim = int(case["head_dim"])
    rotary_dim = min(int(case["rotary_dim"]), head_dim)
    max_sequence_length = int(case["max_sequence_length"])
    theta = float(case["theta"])

    if head_dim % 2 != 0:
        raise ValueError("RoPE head_dim must be even")
    if rotary_dim % 2 != 0:
        raise ValueError("RoPE rotary_dim must be even")

    yarn = parse_rope_scaling(case.get("rope_scaling"), max_sequence_length)
    attention_scaling = float(yarn["attention_factor"]) if yarn else 1.0

    inv = inv_freqs(rotary_dim, theta)
    if yarn is not None:
        apply_yarn_scaling(inv, rotary_dim, theta, yarn)

    values: list[float] = []
    rotary_pairs = rotary_dim // 2
    tail_pairs = head_dim // 2 - rotary_pairs

    for pos in range(max_sequence_length):
        t = float(pos)

        for inv_f in inv:
            angle = t * inv_f
            values.append(math.cos(angle) * attention_scaling)
            values.append(math.sin(angle) * attention_scaling)

        for _ in range(tail_pairs):
            values.append(1.0)
            values.append(0.0)

    return {
        "head_dim": head_dim,
        "rotary_dim": rotary_dim,
        "max_sequence_length": max_sequence_length,
        "theta": theta,
        "attention_scaling": attention_scaling,
        "values": values,
    }


def load_case(path: Path) -> dict[str, Any]:
    with path.open("r", encoding="utf-8") as f:
        case = json.load(f)
    if not isinstance(case, dict):
        raise ValueError("case file must contain a JSON object")
    return case


def main() -> int:
    parser = argparse.ArgumentParser(description="Generate RoPE reference data.")
    parser.add_argument("--case", required=True, type=Path, help="Input case JSON file")
    parser.add_argument(
        "--output",
        type=Path,
        help="Output JSON file. If omitted, write to stdout.",
    )
    args = parser.parse_args()

    case = load_case(args.case)
    result = generate_rotary_embedding(case)
    payload = json.dumps(result, ensure_ascii=False, indent=2, sort_keys=True)

    if args.output is None:
        print(payload)
    else:
        args.output.parent.mkdir(parents=True, exist_ok=True)
        args.output.write_text(payload + "\n", encoding="utf-8")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
